.github/workflows/staging.yml

```diff
@@ -1,12 +1,52 @@
-name: Staging Deployment
+name: Deploy Staging
 
-on: workflow_dispatch
+on:
+  workflow_dispatch:
+    inputs:
+      deployMain:
+        description: "Deploy main? true/false"
+        required: false
+        default: "true"
 
 jobs:
-  terraform_plan:
-    name: Terraform plan
+  save_docker_image_tags_staging:
+    if: github.event.inputs.deployMain == 'true'
+    name: Copy and save image tags for staging environment
     runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          token: ${{ secrets.MCA_GITHUB_TOKEN }}
+      - name: Save image tags for staging environment
+        run: |
+          cd terraform
+          cp dev.images.tfvars staging.images.tfvars
+          git config user.name github-actions
+          git config user.email github-actions@github.com
+          git add .
+          git commit -m "Copy and save image tags for use in staging environment [skip ci]" & git push
 
+  terraform_deploy_staging:
+    if: always()
+    needs: [save_docker_image_tags_staging]
+    name: Terraform Deploy
+    runs-on: ubuntu-latest
+    environment: staging
+    defaults:
+      run:
+        working-directory: terraform
+    env:
+      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+      TF_WORKSPACE: staging
     steps:
-      - name: Terraform plan
-        run: echo "TODO Terraform plan"
+      - uses: actions/checkout@v2
+      - uses: hashicorp/setup-terraform@v1
+      - name: Terraform Init
+        run: terraform init
+      - name: Terraform Validate
+        run: terraform validate
+      - name: Terraform Deploy
```

Comment:
Just more of a note for when we do the production deployment but I was thinking that the CI/CD pipeline for production in terms of jobs would be:

- Promote Docker images from staging to production (optional step) - Essentially the same step we have here. Copying across the Docker images from staging into the production file ready for deployment
- Terraform plan - This would do terraform init, validate and then plan using the var files and env variables. This step would essentially be a safety net to allow us to view what the deployment is going to do before we manually approve it in the next step
- Terraform deploy - This is this step here but we would add [required reviewers](https://docs.github.com/en/actions/reference/environments#:~:text=Use%20required%20reviewers%20to%20require,job%20for%20it%20to%20proceed.) to the production environment.  This would allow us to review the changes that Terraform is about to make before we deploy the changes

Thoughts on this?

Reply:
Oooh I definitely like the `terraform plan` and `apply` being separate!!! Yes!!!